torch/_inductor/codegen/cpp.py

```diff
@@ -1665,23 +1739,59 @@ def gen_kernel(kernel):
+                for loop in loops:
+                    if loop.parallel:
+                        for kernel in loop.get_kernels():
+                            for i, line in enumerate(kernel.stores._lines):
+                                if line in kernel.parallel_reduction_stores.keys():
+                                    kernel.stores._lines[
+                                        i
+                                    ] = kernel.parallel_reduction_stores[line]
+
             def get_reduction_code_buffer(loops, is_suffix=True):
                 for loop in loops:
                     for kernel in loop.get_kernels():
                         if is_suffix:
-                            return kernel.reduction_suffix
+                            suffix = (
+                                (
+                                    kernel.parallel_reduction_suffix
+                                    + kernel.reduction_suffix,
+                                    None,
+                                    None,
+                                )
+                                if loop.parallel
+                                else (kernel.reduction_suffix, None, None)
+                            )
+                            return suffix
                         else:
-                            return kernel.reduction_prefix
-                return None
+                            prefix = (
+                                (
+                                    kernel.reduction_prefix
+                                    + kernel.parallel_reduction_prefix,
+                                    kernel.local_reduction_init,
+                                    kernel.local_reduction_stores,
+                                )
+                                if loop.parallel
+                                else (kernel.reduction_prefix, None, None)
+                            )
+                            return prefix
 
             def gen_loops(loops: List[LoopLevel], in_reduction=False):
                 with contextlib.ExitStack() as stack_outer:
                     if loops:
                         loop = loops[0]
-                        if loop.is_reduction() and not in_reduction:
-                            reduction_prefix = get_reduction_code_buffer(
-                                loops, is_suffix=False
-                            )
+                        if loop.is_reduction and not in_reduction:
+                            replace_parallel_reduction_store(loops)
+                            (
+                                reduction_prefix,
+                                ws_init,
+                                ws_tores,
+                            ) = get_reduction_code_buffer(loops, is_suffix=False)
+                            if ws_init:
+                                assert ws_tores
+                                worksharing.reduction_init = ws_init
```

Comment:
Why adding the reduction related logic into worksharing, not write the reduction_init after https://github.com/pytorch/pytorch/pull/115620/files#diff-5ab7b0235e2076a5fc6629ba0b109208940f5b94f5c13babc3e0f87cf4fcec82R1799?

Reply:
`parallel_reduction_init` is used to init global reduction vectors and `worksharing_reudction_init` is use to init thread local reduction variables. For example
```cpp
float tmp_acc_arr[64];   // parallel_reduction_init
#omp parallel
{
  auto tid = omp_get_num_thread();
  float tmp_acc_local; // worksharing_reudction_init (triggered by worksharing.parallel() for parallel reduction case)
 #omp for
  for(){
  ...
  }
 tmp_acc_arr[tid] = tmp_acc_local; // (triggered by worksharing.close() for parallel reduction case
}
```

Reply:
What's the problem of add to https://github.com/pytorch/pytorch/pull/115620/files#diff-5ab7b0235e2076a5fc6629ba0b109208940f5b94f5c13babc3e0f87cf4fcec82R1799 instead of adding reduction logic into worksharing?

Reply:
The reduction logic in work sharing is to init **thread_local** variables and copy thread_local variables into global ACC arr before they are released. Their life time is between `worksharing.parallel` and `worksharing.close`.
Adding this part to https://github.com/pytorch/pytorch/pull/115620/files#diff-5ab7b0235e2076a5fc6629ba0b109208940f5b94f5c13babc3e0f87cf4fcec82R1799 (outside the worksharing) cannot implement this `thread_local` logic.

Reply:
I mean adding the worksharing init after this
```python
                        if loop_nest.is_reduction_only() and loop.parallel:
                            worksharing.parallel(threads)
```
https://github.com/pytorch/pytorch/pull/115620/files#diff-5ab7b0235e2076a5fc6629ba0b109208940f5b94f5c13babc3e0f87cf4fcec82R1770

Reply:
```
            def gen_loops(loops: List[LoopLevel], in_reduction=False):
                with contextlib.ExitStack() as stack_outer:
                    if loops:
                        loop = loops[0]
                        if loop.is_reduction and not in_reduction:
                            (
                                reduction_prefix,
                                ws_init,
                                ws_tores,
                            ) = get_reduction_code_buffer(loops, is_suffix=False)
                            #if ws_init:
                            #   assert ws_tores
                            #  worksharing.reduction_init = ws_init
                            #   worksharing.reduction_stores = ws_tores
                            if reduction_prefix:
                                stack_outer.enter_context(code.indent())
                            code.splice(reduction_prefix)
                        if loop_nest.is_reduction_only() and loop.parallel:
                            worksharing.parallel(threads)
                            # init thread_local variable here

                    for loop in loops:
                        gen_loop(loop, loops[0].is_reduction and not in_reduction)

                    if loops:
                        loop = loops[0]
                        if loop_nest.is_reduction_only() and loop.parallel:
                            # copy thread_local ACC variable to global arr
                            worksharing.close()
                        if loop.is_reduction and not in_reduction:
                            code.splice(
                                get_reduction_code_buffer(loops, is_suffix=True)[0]
                            )
```

Reply:
I guess the key question is why worksharing has to be aware of "reduction" logic?

Reply:
I refined the code and make `local init/copy` logic out of worksharing.